### PR TITLE
Image recommendations data wiring

### PR DESCRIPTION
**Pull request by sharvaniharan**
**Phab:** https://phabricator.wikimedia.org/T340669

### CODE REVIEW

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class ImageRecommendationsEvent(
+   private val action: String,
+   private val active_interface: String,
+   private val action_data: String,
+   private val primary_language: String,
+   private val wiki_id: String
+) : MobileAppsEvent(STREAM_NAME) {
+
+    companion object {
+        private const val ACTION_IMPRESSION = "impression"
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:18]: Is it necessary to have a `const` for the "impression" string, when we don't have any other constants for any of the other strings?

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            submitImageRecommendationEvent(ACTION_IMPRESSION, activeInterface, "", "")
+        }
+
+        fun logAction(action: String, activeInterface: String, actionData: String, wikiId: String) {
+            submitImageRecommendationEvent(action, activeInterface, actionData, wikiId)
+        }
+
+        fun getActionDataString(filename: String = "", recommendationSource: String = "", recommendationSourceProject: String = "",
+                                rejectionReasons: String = "", acceptanceState: String = "", seriesNumber: String = "",
+                                totalSuggestions: String = "", revisionId: String = "", captionAdd: String = "", altTextAdd: String = ""): String {
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:31]: Each of these string parameters could be nullable (`String? = null`), and the resulting action string doesn't need to include the parameter if it's null.

File: app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt [hunk 3]
```diff
@@ -493,11 +505,27 @@ class EditSectionActivity : BaseActivity(), ThemeChooserDialog.Callback {
     fun clickNextButton() {
         when {
             editSummaryFragment.isActive -> {
+                if (invokeSource == Constants.InvokeSource.EDIT_ADD_IMAGE) {
+                    val pageTitle = intent.getParcelableExtra<PageTitle>(InsertMediaActivity.EXTRA_IMAGE_TITLE)
+                    ImageRecommendationsEvent.logAction("editsummary_save", "editsummary_dialog", ImageRecommendationsEvent.getActionDataString(
+                        filename = pageTitle?.prefixedText!!, recommendationSource = pageTitle.wikiSite.languageCode, recommendationSourceProject = pageTitle.wikiSite.languageCode,
+                        acceptanceState = "accepted", seriesNumber = "", totalSuggestions = "",
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:512]: It's not necessary to pass parameters that are empty (i.e. the same as their default value).

File: app/src/main/java/org/wikipedia/readinglist/ReadingListActivity.kt [hunk 4]
```diff
@@ -33,9 +33,7 @@ class ReadingListActivity : SingleFragmentActivity<ReadingListFragment>() {
 
     override fun onBackPressed() {
         super.onBackPressed()
-        if (intent.getBooleanExtra(EXTRA_READING_LIST_PREVIEW, false)) {
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/readinglist/ReadingListActivity.kt:36]: What's the reason for this change?

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+        fun logImpression(activeInterface: String) {
+            submitImageRecommendationEvent(ACTION_IMPRESSION, activeInterface, "", "")
+        }
+
+        fun logAction(action: String, activeInterface: String, actionData: String, wikiId: String) {
+            submitImageRecommendationEvent(action, activeInterface, actionData, wikiId)
+        }
+
+        fun getActionDataString(filename: String = "", recommendationSource: String = "", recommendationSourceProject: String = "",
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:29]: If the `recommendationSourceProject` is supposed to be similar to the [Growth schema](https://gerrit.wikimedia.org/r/plugins/gitiles/schemas/event/secondary/+/refs/heads/master/jsonschema/analytics/mediawiki/structured_task/article/image_suggestion_interaction/current.yaml#92), then:

- It should be plural `recommendationSourceProjects`
- The value should come from the `GrowthImageSuggestion.ImageItem.projects` list, and _not_ the current language wiki.
- The "projects" list basically says "We're recommending this image because it was also used in x, y, z language wikipedia."

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                                rejectionReasons: String = "", acceptanceState: String = "", seriesNumber: String = "",
+                                totalSuggestions: String = "", revisionId: String = "", captionAdd: String = "", altTextAdd: String = ""): String {
+            return "filename:$filename, recommendation_source:$recommendationSource,recommendation_source_project:$recommendationSourceProject, " +
+                    "rejection_reasons:$rejectionReasons, acceptance_state:$acceptanceState, series_number: $seriesNumber," +
+                    "total_suggestions: $totalSuggestions, revision_id:$revisionId, caption_add: $captionAdd, alt_text_add: $altTextAdd"
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:34]: It looks like `revision_id`, `caption_add`, and `alt_text_add` are not documented in the [Readme](https://gerrit.wikimedia.org/r/plugins/gitiles/schemas/event/secondary/+/refs/heads/master/jsonschema/analytics/mobile_apps/android_image_recommendation_event/README.md) of the schema. On the other hand, the Readme seems to document a `time_spent` parameter, which I'm not seeing here.

Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt]: [Done](https://gerrit.wikimedia.org/r/c/schemas/event/secondary/+/951181)

**Comment by sharvaniharan** [comment]
@cooltey @dbrant This is ready for review

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+@Serializable
+@SerialName("/analytics/mobile_apps/android_image_recommendation_event/1.0.0")
+class ImageRecommendationsEvent(
+   private val action: String,
+   private val active_interface: String,
+   private val action_data: String,
+   private val primary_language: String,
+   private val wiki_id: String
+) : MobileAppsEvent(STREAM_NAME) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:18]: Not sure if the parameters are unused or not. If not, please add `@Suppress("unused")`

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                DescriptionEditActivity.Action.ADD_CAPTION -> logAction("edit_success", "se_add_caption", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                DescriptionEditActivity.Action.TRANSLATE_CAPTION -> logAction("edit_success", "se_translate_caption", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                else -> logAction("edit_success", "se_add_image_tags", getActionDataString(revisionId = revisionId.toString()), wikiId)
+            }
+        }
+
+        fun getActionDataString(filename: String? = null, recommendationSource: String? = null,
+                                recommendationSourceProjects: String? = null, rejectionReasons: String? = null,
+                                acceptanceState: String? = null, revisionId: String? = null, captionAdd: Boolean? = null,
+                                altTextAdd: Boolean? = null, addTimeSpent: Boolean = false): String {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:49]: Would it be possible to have the parameters have a better format since this function has a lot of parameters? For example:
```
                               (filename: String? = null,
                                recommendationSource: String? = null,
                                recommendationSourceProjects: String? = null,
                                rejectionReasons: String? = null,
                                acceptanceState: String? = null,
                                revisionId: String? = null,
                                captionAdd: Boolean? = null,
                                altTextAdd: Boolean? = null,
                                addTimeSpent: Boolean = false)
```

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            return "${filename?.let { "filename: ${URLEncoder.encode(filename, "UTF-8")}, " }.orEmpty()}${recommendationSource?.let { "recommendation_source: $it, " }.orEmpty()}" +
+                    "${recommendationSourceProjects?.let { "recommendation_source_project: $it, " }.orEmpty()}${rejectionReasons?.let { "rejection_reasons: $it, " }.orEmpty()}" +
+                    "${acceptanceState?.let { "acceptance_state: $it, " }.orEmpty()}${revisionId?.let { "revision_id: $it, " }.orEmpty()}" +
+                    "${captionAdd?.let { "caption_add: $it, " }.orEmpty()}${altTextAdd?.let { "alt_text_add: $it, " }.orEmpty()}${if (addTimeSpent) "alt_text_add: " + timer.elapsedMillis.toString() else ""}"
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:54]: Would it be possible to have an extra `val` for the strings to increase the readability? 
```
val filenameStr = filename?.let { "filename: ${URLEncoder.encode(filename, "UTF-8")}, " }.orEmpty()
val recSourceStr = recommendationSource?.let { "recommendation_source: $it, " }.orEmpty()
return filenameStr + recSourceStr
```

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                timer.reset()
+            }
+            submitImageRecommendationEvent("impression", activeInterface, actionData, wikiId)
+        }
+
+        fun logAction(action: String, activeInterface: String, actionData: String = "", wikiId: String = "") {
+            submitImageRecommendationEvent(action, activeInterface, actionData, wikiId)
+        }
+
+        fun logSeEditSuccess(action: DescriptionEditActivity.Action, wikiId: String, revisionId: Long) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:36]: Can we name the function `logEditSuccess` since this function is under `ImageRecommendationEvent` class?

File: app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt [hunk 11]
```diff
@@ -494,13 +505,31 @@ class EditSectionActivity : BaseActivity(), ThemeChooserDialog.Callback {
      * based on the current state of the button.
      */
     fun clickNextButton() {
+        val addImageTitle = intent.getParcelableExtra<PageTitle>(InsertMediaActivity.EXTRA_IMAGE_TITLE)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:508]: We can use `parcelableExtra` here.

File: app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt [hunk 12]
```diff
@@ -494,13 +505,31 @@ class EditSectionActivity : BaseActivity(), ThemeChooserDialog.Callback {
...
+        val addImageSource = intent.getStringExtra(InsertMediaActivity.EXTRA_IMAGE_SOURCE)
         when {
             editSummaryFragment.isActive -> {
+                if (invokeSource == Constants.InvokeSource.EDIT_ADD_IMAGE) {
+                    ImageRecommendationsEvent.logAction("editsummary_save", "editsummary_dialog", ImageRecommendationsEvent.getActionDataString(
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:513]: Do you think it would be better if we put these log actions into `ImageRecommendationsEvent`?

For example: 
```
fun logEditSummaryAction(actionData: String = "", wikiId: String = "") {
        logAction("editsummary_save", "editsummary_dialog", actionData, wikiId)
}
```

File: app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt [hunk 13]
```diff
@@ -754,6 +789,8 @@ class EditSectionActivity : BaseActivity(), ThemeChooserDialog.Callback {
     }
 
     override fun onBackPressed() {
+        val addImageTitle = intent.getParcelableExtra<PageTitle>(InsertMediaActivity.EXTRA_IMAGE_TITLE)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:792]: We can use `parcelableExtra` here

File: app/src/main/java/org/wikipedia/edit/insertmedia/InsertMediaActivity.kt [hunk 14]
```diff
@@ -152,10 +153,23 @@ class InsertMediaActivity : BaseActivity() {
                 true
             }
             R.id.menu_save -> {
+                if (viewModel.invokeSource == Constants.InvokeSource.EDIT_ADD_IMAGE && viewModel.selectedImage != null) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/edit/insertmedia/InsertMediaActivity.kt:156]: Can we assign a value for `viewModel.selectedImage` before the if statement so that we do not have to put `?` or `!!` for the parameters?
`val selectedImage = viewModel.selectedImage`

File: app/src/main/java/org/wikipedia/suggestededits/SuggestedEditsImageRecsFragment.kt [hunk 15]
```diff
@@ -367,9 +425,33 @@ class SuggestedEditsImageRecsFragment : SuggestedEditsItemFragment(), MenuProvid
...
         viewModel.rejectRecommendation(null, selectedItems)
+        ImageRecommendationsEvent.logAction("reject_submit", "rejection_dialog",
+            getActionStringForAnalytics(rejectionReasons = selectedItems.map { ImageRecommendationsEvent.reasons.getOrNull(it) }.toString(),
+            acceptanceState = "rejected"), viewModel.langCode)
         publish()
     }
 
+    private fun getActionStringForAnalytics(acceptanceState: String? = null, rejectionReasons: String? = null,
+                                            revisionId: String? = null, addTimeSpent: Boolean = false): String {
+        return ImageRecommendationsEvent.getActionDataString(filename = viewModel.recommendation.images[0].image,
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/suggestededits/SuggestedEditsImageRecsFragment.kt:436]: Can we have `val firstImage = viewModel.recommendation.images[0]` here?

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        fun getActionDataString(filename: String? = null, recommendationSource: String? = null,
+                                recommendationSourceProjects: String? = null, rejectionReasons: String? = null,
+                                acceptanceState: String? = null, revisionId: String? = null, captionAdd: Boolean? = null,
+                                altTextAdd: Boolean? = null, addTimeSpent: Boolean = false): String {
+            return "${filename?.let { "filename: ${URLEncoder.encode(filename, "UTF-8")}, " }.orEmpty()}${recommendationSource?.let { "recommendation_source: $it, " }.orEmpty()}" +
+                    "${recommendationSourceProjects?.let { "recommendation_source_project: $it, " }.orEmpty()}${rejectionReasons?.let { "rejection_reasons: $it, " }.orEmpty()}" +
+                    "${acceptanceState?.let { "acceptance_state: $it, " }.orEmpty()}${revisionId?.let { "revision_id: $it, " }.orEmpty()}" +
+                    "${captionAdd?.let { "caption_add: $it, " }.orEmpty()}${altTextAdd?.let { "alt_text_add: $it, " }.orEmpty()}${if (addTimeSpent) "alt_text_add: " + timer.elapsedMillis.toString() else ""}"
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:53]: Incorrect parameter name for `addTimeSpent`.

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            submitImageRecommendationEvent(action, activeInterface, actionData, wikiId)
+        }
+
+        fun logSeEditSuccess(action: DescriptionEditActivity.Action, wikiId: String, revisionId: Long) {
+            when (action) {
+                DescriptionEditActivity.Action.ADD_DESCRIPTION -> logAction("edit_success", "se_add_description", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                DescriptionEditActivity.Action.TRANSLATE_DESCRIPTION -> logAction("edit_success", "se_translate_description", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                DescriptionEditActivity.Action.ADD_CAPTION -> logAction("edit_success", "se_add_caption", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                DescriptionEditActivity.Action.TRANSLATE_CAPTION -> logAction("edit_success", "se_translate_caption", getActionDataString(revisionId = revisionId.toString()), wikiId)
+                else -> logAction("edit_success", "se_add_image_tags", getActionDataString(revisionId = revisionId.toString()), wikiId)
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:42]: Do we not want to log `edit_success` for image recommendations?

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        }
+
+        fun getActionDataString(filename: String? = null, recommendationSource: String? = null,
+                                recommendationSourceProjects: String? = null, rejectionReasons: String? = null,
+                                acceptanceState: String? = null, revisionId: String? = null, captionAdd: Boolean? = null,
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:48]: `revisionId` can be `Long`.

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+                                altTextAdd: Boolean? = null, addTimeSpent: Boolean = false): String {
+            return "${filename?.let { "filename: ${URLEncoder.encode(filename, "UTF-8")}, " }.orEmpty()}${recommendationSource?.let { "recommendation_source: $it, " }.orEmpty()}" +
```
Comment (dbrant) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:50]: We have `UriUtil.encodeURL()`.

**Review by dbrant — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt (hunk 17 above)
Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:42]: We decided we can have just one event for both `success` and `success_cofirm` [to ensure feedback snackbar was shown] here : https://github.com/wikimedia/apps-android-wikipedia/pull/4092/files#diff-af6debb3eecb7f1791753ccdfa9a9337be4d28f3d6d026b89774aa2cb4a24ce2R83

File: app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt (hunk 12 above)
Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:513]: `action` string is the one that changes per event, so i feel this wont be useful.

Comment (cooltey) [reply app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt:513]: Got it. It makes sense to me.

File: app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt [hunk 20]
```diff
@@ -33,24 +34,38 @@ class ImageRecommendationsEvent(
...
+        ): String {
+            val filenameStr = filename?.let { "filename: ${UriUtil.encodeURL(filename)}, " }.orEmpty()
+            val recSourceStr = recommendationSource?.let { "recommendation_source: $it, " }.orEmpty()
+            val recSourceProjectsStr = recommendationSourceProjects?.let { "recommendation_source_project: $it, " }.orEmpty()
+            val rejectionReasonsStr = rejectionReasons?.let { "rejection_reasons: $it, " }.orEmpty()
+            val acceptanceStateStr = acceptanceState?.let { "acceptance_state: $it, " }.orEmpty()
+            val revisionIdStr = revisionId?.let { "revision_id: $it, " }.orEmpty()
+            val captionAddStr = captionAdd?.let { "caption_add: $it, " }.orEmpty()
+            val altTextAddStr = altTextAdd?.let { "alt_text_add: $it, " }.orEmpty()
+            val timeSpentStr = if (addTimeSpent) "time_spent: " + timer.elapsedMillis.toString() else ""
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/eventplatform/ImageRecommendationsEvent.kt:66]: Would it be simpler and clearer if we changed to:
```
if (addTimeSpent) "time_spent: ${timer.elapsedMillis}" else ""
```

**Review by cooltey — CHANGES_REQUESTED** [review]
Looks good to me and there is one last comment in the event class.
